### ISSUE TITLE
Feature Request: Preconditioned Matrix Optimizers

### ISSUE THREAD

**Issue by plugyawn** [NVIDIA/Megatron-LM#5345]
**Is your feature request related to a problem? Please describe.**
Preconditioned SGD is a powerful family of optimizers, including: FOOF, KFAC, new optimizers like Newton-Muon (which can be seen as Muon msign( ) over FOOF). Recently, a number of prominent names in optimization theory have called for PSGD to be scaled as well. 

Tag @NVIDIA/mcore-oncall
to get oncall's attention to this issue.

**Describe the solution you'd like**

PSGD usually suffers from the problem of often having to communicate activation statistics: FOOF uses the input feature gram, KFAC uses both input feature gram and the gradient gram matrix. However, we can show that this is avoidable by calculating it with the `w_grad` inside TransformerEngine: `w_grad = dY.T @ X`,  `feature_gram  = X.T @ X` and `grad_gram = dY.T @ dY`. 

In fact, some more obscure optimizers like the simple variant of Local Loss propagation (Locoprop-S) can be expressed as preconditioned gradient descent with these building blocks as well. 

It is also straightforward, then, to calculate MuP parameterization for these systems, unlike AdamW, where the weight decay is a problem. 

For FSDP, instead of the full preconditioning matrices, we may show that a diagonal/block-diagonal approximation often suffices, and imo can be cleanly implemented after we have FSDP support for Muon (#3179). 

I have a reference implementation up at: https://github.com/plugyawn/Megaprop.

<img width="2536" height="1948" alt="Image" src="https://github.com/user-attachments/assets/e7b961c2-6e5e-40a6-a446-6ac36ca54a72" />

<img width="3252" height="1930" alt="Image" src="https://github.com/user-attachments/assets/21485119-fd5b-421d-a594-8ef97282fa06" />

Megaprop pins TransformerEngine, Megatron-LM, as well as Emerging-Optimizers. It has been verified at 2/4/8 GPUs so far, but unfortunately I don't have the resources to check further.

@sbhavani not sure who the on-call is, so tagging since we had a small discussion about this. 
Same: @mkhona-nvidia.

**Comment by guihong-nv** [comment]
Thanks for the detailed feature request and for sharing the reference implementation.

This is an interesting direction, and we really appreciate the work you’ve put into exploring preconditioned matrix optimizers in the Megatron-LM stack. For now, we do not plan to add official support for this, mainly because it has not yet been validated at large-scale training setup.

That said, we highly appreciate the contribution and the discussion here. We would be happy to see more community experimentation, benchmarks, and PRs in this area, especially results demonstrating stability and performance at larger scale.

**Comment by plugyawn** [comment]
That makes sense! I imagine it would take some time for large-scale validation, although Newton-Muon is very promising. 

The MuP direction is already underway in Megatron: #3058, #3715, #4381, #4998, #4999, #5000.

Here are some more results on width MuP upto 4096 and depth-MuP upto 32 from base depth 2.
<img width="1100" height="975" alt="Image" src="https://github.com/user-attachments/assets/571bfbfa-3ad0-4771-94a1-35c876a6572a" />

<img width="1260" height="981" alt="Image" src="https://github.com/user-attachments/assets/fa046a65-24b2-42c7-9bf6-9b712331d076" />

**Comment by ericharper** [comment]
@skyw any feedback on this Feature Request?

**Comment by skyw** [comment]
> [@skyw](https://github.com/skyw) any feedback on this Feature Request?

I think the request need to be breakdown to individual items. "Preconditioned Matrix Optimizers" is very wide.

**Comment by plugyawn** [comment]
> I think the request need to be breakdown to individual items. "Preconditioned Matrix Optimizers" is very wide.

@skyw @ericharper, I think [Megaprop](https://github.com/plugyawn/megaprop) can be broken roughly into the following logical PRs:

1. TE support for `feature_gram` and `grad_gram` sidecars.
2. Megatron-LM support for FOOF by consuming `feature_gram` from TE and right-preconditoning with MuP support.
3. Support for Newton-Muon, which is FOOF + Muon, implementation-wise.
4. KFAC support / Emerging-Optimizers support for extra utils as and where needed.

While "preconditioned matrix optimizer" is very wide indeed, most standard preconditioning techniques boil down to some combination of the `feature_gram` and the `grad_gram` applied at the gradient step -- for example, NM becomes trivial after 1 and 2 are implemented. 

Megaprop also has a Muon ÷FSDP implementation, which could also prove useful, but I haven't verified it in >8 GPU/cross-node situation. Megaprop doesn't alias any of these optimizers by name (i.e, no mention of KFAC/NM/FOOF, etc), which is awful for keeping track of what is what. 

Some more details on [this link](https://x.com/plugyawn/status/2066558223113691242), but from the same:

For eg: right-prec. with the feature gram retrieves FOOF.
Composing with Muon retrieves Newton-Muon.

For Newton Muon, in Megaprop, we pass:
--matrix-optimizer muon --matrix-input-preconditioner feature_gram

If you want a diagonal approximation of the feature_gram, we also pass.
--matrix-input-preconditioner-approximation diag
